tools/Dashboard/DevHome.Dashboard/Views/DashboardView.xaml.cs

```diff
@@ -514,6 +518,67 @@ private async Task EditWidget(WidgetViewModel widgetViewModel)
         widgetViewModel.IsInEditMode = false;
     }
 
+    private void WidgetGridView_DragItemsStarting(object sender, DragItemsStartingEventArgs e)
+    {
+        Log.Logger()?.ReportDebug("DashboardView", $"Drag starting");
+
+        // When drag starts, save the WidgetViewModel and the original index of the widget being dragged.
+        var draggedObject = e.Items.FirstOrDefault();
+        var draggedWidgetViewModel = draggedObject as WidgetViewModel;
+        e.Data.Properties.Add(DraggedWidget, draggedWidgetViewModel);
+        e.Data.Properties.Add(DraggedIndex, PinnedWidgets.IndexOf(draggedWidgetViewModel));
+    }
+
+    private void WidgetControl_DragOver(object sender, DragEventArgs e)
+    {
+        // A widget may be dropped on top of another widget, in which case the dropped widget will take the target widget's place.
+        e.AcceptedOperation = Windows.ApplicationModel.DataTransfer.DataPackageOperation.Move;
+    }
+
+    private async void WidgetControl_Drop(object sender, DragEventArgs e)
+    {
+        Log.Logger()?.ReportDebug("DashboardView", $"Drop starting");
+
+        // When drop happens, get the original index of the widget that was dragged and dropped.
+        e.Data.Properties.TryGetValue(DraggedIndex, out var draggedIndexObject);
+        var draggedIndex = (int)draggedIndexObject;

```

Comment:
what happens if the drag starts in another app? likely need to error out if we don't have a valid set of properties rather than assume we do :)